New tutorial for payment channel example (#67)
hyperledger-labs/perun-doc: source/go-perun/new_tutorial/main/setup.rst

.. _setupscenario:

Setup Scenario
##############
Before creating our exemplary scenario, we construct a few helper functions that will cover the contract deployment, Client generation, and logging of balances in the command line.
We put the code of this section into `util.go`

Deploy Contracts
================

go-perun uses two on-chain contracts: the Adjudicator and the Asset Holder.

> They are written in the contract language of Ethereum; solidity.

Reviewer: Could add a link to the solidity docs.

> They are written in the contract language of Ethereum; solidity.

Reviewer: ```suggestion
They are written in the contract language of Ethereum, Solidity, and are part of go-perun's Ethereum backend.
```

Each contract must be deployed before go-perun can be used.
Usually, you would assume that they are already deployed, and the addresses are known in advance.
But since this is a complete example for a local chain, we must deploy them.

Concept: Adjudicator  & Asset Holder
------------------------------------
The Adjudicator contract ensures that a user can consistently enforce the rules of his Channel.

> Since the central part of the communication is off-chain, he would only contact the Adjudicator if he felt betrayed by the participants.

Reviewer: ```suggestion
Since the central part of the communication is off-chain, he would only contact the Adjudicator if he felt betrayed by the participants or for concluding the channel.
```

The Asset Holder holds the on-chain balances for all Ledger Channels. It is always associated with a specific Adjudicator instance.

All participants deposit their funds into the Asset Holder before a Channel can be opened.
All participants can withdraw their funds via the Asset Holder when the Channel is closed.
In the case of a dispute, the Asset Holder respects the decision of its Adjudicator on how to proceed.

go-perun uses one contract per asset on top of the Ethereum blockchain.
In this example, we only use the `ETHAssetHolder`, which is used for Ether, the native currency in Ethereum.
ERC20 Tokens are supported via the `ERC20AssetHolder`.

Implementation
--------------

Go-perun makes the deployment of the standard Adjudicator and Ethereum AssetHolder easy.
As stated before: In a running go-perun ecosystem, the contracts' addresses would be known in advance, and you would verify them.

> We will deploy both contracts here for demonstration:

Reviewer: ```suggestion
In our example, we will deploy contracts ourselves at the start of the program:
```

            panic(err)
        }
        w := swallet.NewWallet(k)
        cb, err := client.CreateContractBackend(nodeURL, chainID, w)
        if err != nil {
            panic(err)
        }
        acc := accounts.Account{Address: crypto.PubkeyToAddress(k.PublicKey)}

Using the Contract Backend `cb`, we then deploy the Adjudicator and Ethereum AssetHolder via go-perun's `ethchannel.DeployAdjudicator()` and `ethchannel.DeployETHAssetholder()`.
Note that the Adjudicator must be deployed first for using its address to make the AssetHolder depend on it.
Ultimately we return both addresses.

.. code-block:: go

        // Deploy adjudicator.
        adj, err = ethchannel.DeployAdjudicator(context.TODO(), cb, acc) //TODO:go-perun accept ethwallet Account instead?
        if err != nil {
            panic(err)
        }

        // Deploy asset holder.
        ah, err = ethchannel.DeployETHAssetholder(context.TODO(), cb, adj, acc) //TODO:go-perun accept ethwallet Account instead?
        if err != nil {
            panic(err)
        }

        return adj, ah
    }

Nice-To-Haves
=============
The following functions will help us generate or scenario in a neat way.

> Client generation by the private key

Reviewer: ```suggestion
Client setup from secret key
```
